### PR TITLE
Group entity

### PR DESCRIPTION
**Pull request by zugnachpankow**
I think the group entity would be at a point to be merged into master. I only ran tests that used the entity with a modified seven dwarfs script, but i also ran exploit (w/o use of group entity) and had no errors due to the implementation of the group entity in the global data structure and the base component. If any problems arise with other studies or whatsoever would be good to know early on. 

Changes:
- implement group entity as globally available class
- enable overlapping memberships of individuals in groups via a directed network (props to @mensch72 for this idea)
- adjust implementation

### CODE REVIEW

**Comment by zugnachpankow** [comment]
Add some further test: full technical functionality should be given now

File: docs/framework_documentation/abstract_level/entity_types/group.rst [hunk 1]
```diff
@@ -33,25 +33,30 @@ Basic relationships to other entity-types
...
--  have several member :doc:`individuals<individual>`
+-  have several members :doc:`individuals<individual>`,
+   which is represented by a group membership directed network owned by the culture taxon
 
 In addition, a group may...
 
--  have one or several "leader" :doc:`individuals<individual>`, 
-   of which one may be the dominant leader
+-  have an "intra" group network between members :doc:`individuals<individual>`
+   (this feature is not completely implemented yet)
```
Comment (jnnsbrr) [inline docs/framework_documentation/abstract_level/entity_types/group.rst:42]: only describe (and explicitly) what feature is implemented and add `TODO` tags in the code.

File: docs/framework_documentation/abstract_level/entity_types/group.rst [hunk 2]
```diff
@@ -33,25 +33,30 @@ Basic relationships to other entity-types
...
--  have a "headquarters" :doc:`cell<cell>`
+-  have one or several "leader" :doc:`individuals<individual>`,
+   of which one may be the dominant leader (this feature is not implemented yet)
 
--  be related to other groups via some network owned by the culture taxon 
+-  have a "headquarters" :doc:`cell<cell>` (this feature is not implemented yet)
+
+-  be related to other groups via an "inter" network owned by the culture taxon
    (which will typically interact with the network of personal acquaintance between member individuals)
+   - (this feature is not completely implemented yet)
```
Comment (jnnsbrr) [inline docs/framework_documentation/abstract_level/entity_types/group.rst:51]: same here.

File: docs/framework_documentation/abstract_level/entity_types/group.rst [hunk 3]
```diff
@@ -33,25 +33,30 @@ Basic relationships to other entity-types
...
 
--  act as the current "elite" in some :doc:`social system<social system>`
+-  act as the current "elite" in some :doc:`social system<social system>` (this feature is not implemented yet)
```
Comment (jnnsbrr) [inline docs/framework_documentation/abstract_level/entity_types/group.rst:53]: and here.

File: docs/framework_documentation/abstract_level/entity_types/group.rst [hunk 4]
```diff
@@ -33,25 +33,30 @@ Basic relationships to other entity-types
...
 
 All these relationships may be dynamic.
 
 Finally, a group may...
 
 -  be a permanent subgroup of a larger group or :doc:`social system<social system>` *by definition* 
    (rather than by coincidence, e.g., "scientists" are by definition a subgroup of the group "academics",
-   and "German workers" may be by definition a subgroup of the social system "Germany")
+   and "German workers" may be by definition a subgroup of the social system "Germany") - (this feature is not implemented yet)
```
Comment (jnnsbrr) [inline docs/framework_documentation/abstract_level/entity_types/group.rst:61]: and here.

File: pycopancore/model_components/base/implementation/group.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        """
+        super().__init__(**kwargs)  # must be the first line
+
+        # init and set variables implemented via properties:
+        self._culture = None
+        self.culture = culture
+        self._world = None
+        self.world = world
+
+        # self._social_system = None
```
Comment (jnnsbrr) [inline pycopancore/model_components/base/implementation/group.py:59]: remove code in comments

File: pycopancore/model_components/base/implementation/group.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    @property  # read-only
+    def metabolism(self):
+        """Get the Metabolism of which the Group is a part."""
+        return self._world.metabolism
+
+    # _higher_groups = unknown
+    """cache, depends on self.next_higher_group
+    and self.next_higher_group.higher_groups"""
+    # @property  # read-only
+    # def higher_groups(self):
```
Comment (jnnsbrr) [inline pycopancore/model_components/base/implementation/group.py:181]: delete commented out code

File: pycopancore/model_components/base/implementation/group.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    def world(self, w):
+        """Set the World the Group is part of."""
+        if self._world is not None:
+            # first deregister from previous world's list of cells:
+            self._world.groups.remove(self)
+        assert isinstance(w, I.World), "world must be of entity type World"
+        w._groups.add(self)
+        self._world = w
+
+    # @property
```
Comment (jnnsbrr) [inline pycopancore/model_components/base/implementation/group.py:129]: delete commented out code.

File: pycopancore/model_components/base/interface.py [hunk 8]
```diff
@@ -288,3 +298,56 @@ class Individual (object, metaclass=_MixinType):
...
+
+
+    # read-only attributes storing redundant information:
+    environment = ReferenceVariable("environment", "", type=Environment,
+                               readonly=True)
+    metabolism = ReferenceVariable("metabolism", "", type=Metabolism,
+                                   readonly=True)
+    culture = ReferenceVariable("culture", "", type=Culture,
+                                readonly=True)
+    # higher_groups = SetVariable(
```
Comment (jnnsbrr) [inline pycopancore/model_components/base/interface.py:324]: remove commented out code

File: pycopancore/model_components/seven_dwarfs/implementation/group.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+from ...base import interface as B
+
+# TODO: import those process types you need:
+from .... import Explicit
+
+class Group (I.Group):
+    """Group entity type mixin implementation class."""
+
+    # standard methods:
+    # TODO: only uncomment when adding custom code!
```
Comment (jnnsbrr) [inline pycopancore/model_components/seven_dwarfs/implementation/group.py:29]: remove commented out code.

**Review by jnnsbrr — COMMENTED** [review]
Hey @zugnachpankow, I did not find any major problems in the code, just some minor style related things.
Fine from my side 🙂

**Review by jnnsbrr — DISMISSED** [review]

**Review by jnnsbrr — APPROVED** [review]
Great, thanks a lot!

**Pull request merged**